client-go/features: introduce feature gates

p0lyn0mial opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?
/kind feature
<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:
This PR add a feature gates mechanism to client-go as described in https://docs.google.com/document/d/1g9BGCRw-7ucUxO6OtCWbb3lfzUGA_uU9178wLdXAIfs

In particular:
  - Adds a default feature gate implementation based on environment variables.
  - Adds a set of methods for reading, overwriting the default implementation, and adding features to an external registry.
#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->

xref: https://github.com/kubernetes/enhancements/issues/3157

#### Special notes for your reviewer:

This is just a first PR. Ben and I will be adding more once this PR merges.


#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Introduce a feature gate mechanism to client-go. Depending on the actual implementation, users can control features via environmental variables or command line options.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs
- [Other doc]: https://docs.google.com/document/d/1g9BGCRw-7ucUxO6OtCWbb3lfzUGA_uU9178wLdXAIfs/edit#heading=h.lymszxjovg65
```

p0lyn0mial committed:
client-go/features: introduce feature gates

This PR add a feature gates mechanisim to client-go
as described in https://docs.google.com/document/d/1g9BGCRw-7ucUxO6OtCWbb3lfzUGA_uU9178wLdXAIfs

In particular:
  - Adds a default feature gate implementation based on environment variables.
  - Adds a set of methods for reading, overwriting the default implementation, and adding features to an external registry.

Co-authored-by: deads2k <deads@redhat.com>
Co-authored-by: Ben Luddy <bluddy@redhat.com>

p0lyn0mial commented:
/assing @benluddy

alexzielenski commented:
/assign @benluddy 
/triage accepted

benluddy commented on staging/src/k8s.io/client-go/features/env_var_feature_gate_test.go:
```diff
@@ -0,0 +1,153 @@
+	}
+	for _, scenario := range scenarios {
+		t.Run(scenario.name, func(t *testing.T) {
+			for k, v := range scenario.envVariables {
+				os.Setenv(k, v)
+			}
+			defer func() {
+				for k := range scenario.envVariables {
+					os.Unsetenv(k)
+				}
+			}()
```
https://pkg.go.dev/testing#T.Setenv could be appropriate here.

p0lyn0mial replied:
yes, thanks!

p0lyn0mial commented on staging/src/k8s.io/client-go/features/features.go:
```diff
@@ -0,0 +1,126 @@
+	featureGates.Store(wrappedFeatureGate)
+}
+
+// featureGatesWrapper a thin wrapper to satisfy featureGates variable (atomic.Value)
+type featureGatesWrapper struct {
+	Reader
+}
```
sure, np

p0lyn0mial commented on staging/src/k8s.io/client-go/features/features.go:
```diff
@@ -0,0 +1,126 @@
+// Please note that the actual set of the features gates
+// might be overwritten by calling SetFeatureGates method.
+func FeatureGates() Reader {
+	return featureGates.Load().(Reader)
```
yes, this is a good idea, thanks.

p0lyn0mial commented on staging/src/k8s.io/client-go/features/testing/features_init_test.go:
```diff
@@ -0,0 +1,43 @@
+	"k8s.io/client-go/features"
+)
+
+func TestDriveInitDefaultFeatureGates(t *testing.T) {
```
I will slightly extend the test to check both, the type and the behaviour.

p0lyn0mial commented:
/assign @sttts

p0lyn0mial commented on staging/src/k8s.io/client-go/features/env_var_feature_gate.go:
```diff
@@ -0,0 +1,121 @@
+			}
+			boolVal, boolErr := strconv.ParseBool(featureState)
+			switch {
+			case boolErr != nil:
+				utilruntime.HandleError(fmt.Errorf("cannot set feature gate %v to %v, due to %v", feature, featureState, boolErr))
+			case featureSpec.LockToDefault:
+				if boolVal != featureSpec.Default {
+					utilruntime.HandleError(fmt.Errorf("cannot set feature gate %v to %v, feature is locked to %v", feature, featureState, featureSpec.Default))
+					break
+				}
```
The main purpose of this function is to give admins a way to disable faulty behaviour. The primary users are kube components which will expose client-go's gates as command line options (potential parsing errors will be reported).  Possibly in a phased manner, initially a feature will be enabled only for a limited number of components. Hopefully this will give developers enough time to test their features  and remove the gates altogether.

For components that wish to test a feature early and do not expose flags as command line options, any potential parsing errors will be logged into a log file, and the default value will be applied. We could change the way we report parsing errors by:
-  crashing the entire program on errors but this approach seems to be way too powerful.
-  exposing a new method that would return any errors and rely on binaries to make use of the new function.

Note that the current behaviour is similar when users set `KUBE_CLIENT_BACKOFF_BASE` and `KUBE_CLIENT_BACKOFF_DURATION`
https://github.com/kubernetes/kubernetes/blob/78cb3862f11225135afdf76f3424e2d7b33104c7/staging/src/k8s.io/client-go/rest/client.go#L148

benluddy replied:
I'm hesitant to suggest a panic too, but it has to be weighed against the scenario where a severe bug is discovered in a default-enabled client feature (say for example a bug that can result in data loss/corruption). If an administrator tries to disable the feature by setting the environment variable `KUBE_FEATURE_Foo=fales` instead of the correct `KUBE_FEATURE_Foo=false`, we can be reasonably sure that their intention was to set an override for the feature named Foo, and they are likely to believe that they have successfully mitigated the bug when in fact nothing has changed and the impact of the bug is prolonged.

Sorry for equivocating. I think the additional logging you suggested is a good idea and provides a way for users to verify that feature gate states are as they expect in non-kube components.

p0lyn0mial replied:
Since client-go is a library, our options are limited, assuming we would like to auto-detect parsing issues.

Another option that came to my mind would be to slightly extend the `Reader` interface to:
```
type Reader interface {
    Enabled(string) (bool, error)
}
```

The additional error could be used to report potential issues, but this solution wouldn't report parsing issues in a clear way, so I don't like this approach.

I also think that checking the additional logging for 3rd party components is a reasonable ask.

benluddy replied:
> I also think that checking the additional logging for 3rd party components is a reasonable ask.

Agreed. I think the behavior you have here is reasonable given the constraints that come with being a library, especially if it's trivial for administrators to check the effective feature state.

p0lyn0mial commented on staging/src/k8s.io/client-go/features/env_var_feature_gate.go:
```diff
@@ -0,0 +1,121 @@
+func (f *envVarFeatureGate) getEnabledMapFromEnvVar() map[Feature]bool {
+	f.readEnvVarsOnce.Do(func() {
+		featureGateState := map[Feature]bool{}
+		for feature, featureSpec := range f.known {
```
how about logging state of every know feature here ?

benluddy replied:
SGTM since there is already precedent for logging directly from client-go.

p0lyn0mial commented:
/retest

benluddy commented:
/lgtm

sttts commented on staging/src/k8s.io/client-go/features/OWNERS:
```diff
@@ -0,0 +1,4 @@
+# See the OWNERS docs at https://go.k8s.io/owners
+
+approvers:
+  - feature-approvers
```
that's a large number of ppl. Would not expect many of them owning features here.

p0lyn0mial replied:
would you prefer to remove this file and rely on the top-level one ?

sttts replied:
yes, between the two options the client-go approvers is the better set of ppl.

p0lyn0mial replied:
done.

sttts commented on staging/src/k8s.io/client-go/features/env_var_feature_gate.go:
```diff
@@ -0,0 +1,130 @@
+// environment variables. The environment variable read for a given feature is formed by
+// concatenating the prefix "KUBE_FEATURE_" with the feature's name.
+//
+// For example, if you have a feature named "MyFeature,"
```
, ?

sttts commented on staging/src/k8s.io/client-go/features/env_var_feature_gate.go:
```diff
@@ -0,0 +1,130 @@
+// setting an environmental variable "KUBE_FEATURE_MyFeature"
+// will allow you to configure the state of that feature.
+//
+// Please note that environmental variables can only be set to a boolean type.
```
what's a boolean?

p0lyn0mial replied:
will change to `the boolean value`.

sttts commented on staging/src/k8s.io/client-go/features/env_var_feature_gate.go:
```diff
@@ -0,0 +1,130 @@
+type envVarFeatureGate struct {
+	// envVarFeatureGateName holds the name of the file
+	// that created this instance
+	envVarFeatureGateName string
```
this stutters. Why the prefix?

sttts replied:
why not say what it is: sourceFileName ?

p0lyn0mial replied:
right, will remove the prefix.

sttts commented on staging/src/k8s.io/client-go/features/features.go:
```diff
@@ -0,0 +1,127 @@
+// By default, only the default features gate will be returned.
+// The default implementation allows controlling the features
+// via environmental variables.
+// For example, if you have a feature named "MyFeature,"
```
comma

sttts commented on staging/src/k8s.io/client-go/features/features.go:
```diff
@@ -0,0 +1,127 @@
+	return registry.Add(defaultKubernetesFeatureGates)
+}
+
+// SetFeatureGates overwrites the default implementation of the feature gate
```
is it one or multiple FeatureGates? I think you mix plural and singnular within this package.

sttts commented on staging/src/k8s.io/client-go/features/features.go:
```diff
@@ -0,0 +1,127 @@
+//	clientgofeaturegate.AddFeaturesToExistingFeatureGates(utilfeature.DefaultMutableFeatureGate)
+//	// then replace client-go's feature gate implementation with your implementation
+//	clientgofeaturegate.SetFeatureGates(utilfeature.DefaultMutableFeatureGate)
+func SetFeatureGates(newFeatureGates Reader) {
```
confusing that the interface is called Reader, but you set the feature gates. `type Gates interface` and `ReplaceGates` or so would be clearer.

`

sttts replied:
How about a mechanism to warn when the EnvVar variant is queried before this is called? That seems to be a bug users could experience that is super hard to track down.

p0lyn0mial replied:
> confusing that the interface is called Reader, but you set the feature gates. type Gates interface and ReplaceGates or so would be clearer.

ok, will rename the interface to `Gates`.

p0lyn0mial replied:
> How about a mechanism to warn when the EnvVar variant is queried before this is called? That seems to be a bug users could experience that is super hard to track down.

sure, i like it, I will open a new pr for this change.

p0lyn0mial replied:
created https://github.com/kubernetes/kubernetes/pull/122641

sttts commented on staging/src/k8s.io/client-go/features/env_var_feature_gates.go:
```diff
@@ -0,0 +1,131 @@
+/*
```
nit: long file name. Why not just envvar.go?

sttts commented on staging/src/k8s.io/client-go/features/env_var_feature_gates.go:
```diff
@@ -0,0 +1,131 @@
+				utilruntime.HandleError(fmt.Errorf("cannot set feature gate %v to %v, due to %v", feature, featureState, boolErr))
+			case featureSpec.LockToDefault:
+				if boolVal != featureSpec.Default {
+					utilruntime.HandleError(fmt.Errorf("cannot set feature gate %v to %v, feature is locked to %v", feature, featureState, featureSpec.Default))
```
nit: %v or %s?

sttts replied:
or even %q

p0lyn0mial replied:
we should go with `%q` as it will be consistent with the `Enabled` method.

sttts commented on staging/src/k8s.io/client-go/features/features.go:
```diff
@@ -0,0 +1,127 @@
+	// LockToDefault indicates that the feature is locked to its default and cannot be changed
+	LockToDefault bool
+	// PreRelease indicates the maturity level of the feature
+	PreRelease prerelease
```
What's our policy to have a private type here? Is that intentional? Doesn't it mean you cannot implement the interface externally?

sttts replied:
Actually, nvmd about implementation. You can create an object and set the value here through the constants.

sttts commented on staging/src/k8s.io/client-go/features/features.go:
```diff
@@ -0,0 +1,127 @@
+
+// Registry represents an external feature gates registry.
+type Registry interface {
+	Add(map[Feature]FeatureSpec) error
```
specify what that does, especially around overriding.

p0lyn0mial commented:
@benluddy @sttts thanks for reviewing this PR. I have applied the last bunch of comments. Please have a look at it and if everything is okay please tag the PR. Once this PR is tagged I'm going to assign it to @deads2k for approval.

p0lyn0mial commented:
I will also start writing up a small blog post explaining the usage of the feature gates in client-go.

deads2k commented:
This lgtm.  I see @sttts has many comments, so I'll leave lgtm with him.

/approve

/assign @sttts

p0lyn0mial committed:
client-go/features: warn when ordering initialization issue

ReplaceFeatureGates logs a warning when the default env var
implementation has been already used.
Such a situation indicates a potential ordering issue and usually is unwanted.

p0lyn0mial commented:
pulled in https://github.com/kubernetes/kubernetes/pull/122641 and addressed the comments.

deads2k commented:
updates addressed Stefan's concerns.  I'm still happy with it

/lgtm

p0lyn0mial commented:
/retest
